### PR TITLE
feat(store): Add store for Releases and Deploys

### PR DESCRIPTION
**Pull request by leedongwei**
Currently, this store fetches `release` and `deploy` separately. However, a `deploy` will always be a child of a `release`. It may be an future optimization target for the enterprise team to combine it into 1 single endpoint.

### CODE REVIEW

File: src/sentry/static/sentry/app/utils/withRelease.tsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          getReleaseStoreKey(projectSlug, releaseVersion)
+        ] = true;
+
+        getReleaseDeploys(api, {orgSlug, projectSlug, releaseVersion});
+      }
+    },
+
+    onStoreUpdate() {
+      const {projectSlug, releaseVersion} = this.props as P & DependentProps;
+      const releaseData = ReleaseStore.get(projectSlug, releaseVersion);
```
Comment (markstory) [inline src/sentry/static/sentry/app/utils/withRelease.tsx:81]: How does state get reset when organizations are switched? While unlikely a user could be in two orgs that have overlapping projects + releases. Should the orgSlug be part of the key used to fetch data from the store, or are we able to rely on that data being overwritten when organizations are changed?

File: src/sentry/static/sentry/app/utils/withRelease.tsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+      if (!releaseData.release && !releaseData.releaseLoading) {
+        // HACK(leedongwei): Actions fired by the ActionCreators are queued to
+        // the back of the event loop, allowing another getRelease for the same
+        // release to be fired before the loading state is updated in store.
+        // This hack short-circuits that and update the state immediately.
+        ReleaseStore.releaseLoading[
+          getReleaseStoreKey(projectSlug, releaseVersion)
+        ] = true;
```
Comment (markstory) [inline src/sentry/static/sentry/app/utils/withRelease.tsx:58]: Should this be done in the actionCreator function instead?

File: src/sentry/static/sentry/app/utils/withRelease.tsx [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    fetchDeploys() {
+      const {api, orgSlug, projectSlug, releaseVersion} = this.props as P &
+        DependentProps;
+      const releaseData = ReleaseStore.get(projectSlug, releaseVersion);
+
+      if (!releaseData.deploys && !releaseData.deploysLoading) {
+        // HACK(leedongwei): Same as above
+        ReleaseStore.deploysLoading[
+          getReleaseStoreKey(projectSlug, releaseVersion)
+        ] = true;
```
Comment (markstory) [inline src/sentry/static/sentry/app/utils/withRelease.tsx:73]: Should this be done in the relevant action creator function?

File: src/sentry/static/sentry/app/actionCreators/releases.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import {Client} from 'app/api';
+import {Deploy, Release} from 'app/types';
+
+type ParamsGet = {
+  orgSlug: string;
+  projectSlug: string;
+  releaseVersion: string;
+};
+
+export function getRelease(api: Client, params: ParamsGet) {
```
Comment (billyvg) [inline src/sentry/static/sentry/app/actionCreators/releases.tsx:13]: Should this be named `getProjectRelease`?

File: src/sentry/static/sentry/app/utils/withRelease.tsx (hunk 1 above)
Comment (billyvg) [reply src/sentry/static/sentry/app/utils/withRelease.tsx:81]: I think you should be able to listen to `'app/actions/organizationActions';` to reset state when orgs are switched

File: src/sentry/static/sentry/app/utils/withRelease.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+      this.fetchDeploys();
+    },
+
+    fetchRelease() {
+      const {api, orgSlug, projectSlug, releaseVersion} = this.props as P &
+        DependentProps;
+      const releaseData = ReleaseStore.get(projectSlug, releaseVersion);
+
+      if (!releaseData.release && !releaseData.releaseLoading) {
+        // HACK(leedongwei): Actions fired by the ActionCreators are queued to
```
Comment (billyvg) [inline src/sentry/static/sentry/app/utils/withRelease.tsx:52]: How/where does the action creator fire twice?

File: src/sentry/static/sentry/app/utils/withRelease.tsx (hunk 2 above)
Comment (leedongwei) [reply src/sentry/static/sentry/app/utils/withRelease.tsx:58]: The comment on Line 52 addresses this.

```
// HACK(leedongwei): Actions fired by the ActionCreators are queued to
// the back of the event loop, allowing another getRelease for the same
// release to be fired before the loading state is updated in store.
// This hack short-circuits that and update the state immediately.
```

Comment (markstory) [reply src/sentry/static/sentry/app/utils/withRelease.tsx:58]: Sure. I'm meant cut/paste this code into the body of the actionCreator function. Right now if we add another callsite for the actioncreator we also need to duplicate this code.

File: src/sentry/static/sentry/app/stores/releaseStore.tsx [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+    orgSlug: undefined,
+    release: new Map() as StoreRelease,
+    releaseLoading: new Map() as StoreLoading,
+    releaseError: new Map() as StoreError,
+    deploys: new Map() as StoreDeploys,
+    deploysLoading: new Map() as StoreLoading,
+    deploysError: new Map() as StoreError,
+  },
+
+  listenables: ReleaseActions,
```
Comment (markstory) [inline src/sentry/static/sentry/app/stores/releaseStore.tsx:56]: Should you also listen to the organization actions to reset state when organizations are changed?

File: src/sentry/static/sentry/app/utils/withRelease.tsx (hunk 5 above)
Comment (leedongwei) [reply src/sentry/static/sentry/app/utils/withRelease.tsx]: If a page uses 2+ components which binds with `withRelease`, then the actionCreator would be fired at least 2+ while rendering the page. This happens on the IssueDetails page.

File: src/sentry/static/sentry/app/stores/releaseStore.tsx (hunk 6 above)
Comment (leedongwei) [reply src/sentry/static/sentry/app/stores/releaseStore.tsx:57]: Great suggestion 😁

**Review by billyvg — APPROVED** [review]

File: tests/js/spec/actionCreators/release.spec.jsx [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+    });
+
+    it('short-circuits the JS event loop when fetching Release', async () => {
+      expect(ReleaseStore.state.releaseLoading[releaseKey]).toEqual(undefined);
+
+      getProjectRelease(api, {orgSlug, projectSlug, releaseVersion});
+      expect(ReleaseActions.loadRelease).toHaveBeenCalled();
+      // expect(ReleaseStore.loadRelease).not.toHaveBeenCalled();
+      expect(ReleaseStore.state.releaseLoading[releaseKey]).toEqual(true);
+    });
```
[the remark this answers is not in the document]
Comment (leedongwei) [reply tests/js/spec/actionCreators/release.spec.jsx:68]: It's tested here at `withRelease`: https://github.com/getsentry/sentry/blob/dc47ecb13b78b033f2bb0bc161c52e481c83c7b4/tests/js/spec/utils/withRelease.spec.jsx#L106

Testing API calls at the HOC-level seems better to me.

[the remark this answers is not in the document]
Comment (billyvg) [reply tests/js/spec/actionCreators/release.spec.jsx:68]: Testing in the actionCreator is closer to where the functionality is, but as long as its covered

File: src/sentry/static/sentry/app/stores/releaseStore.tsx [hunk 8]
```diff
@@ -0,0 +1,227 @@
...
+      deploys: new Map() as StoreDeploys,
+      deploysLoading: new Map() as StoreLoading,
+      deploysError: new Map() as StoreError,
+    };
+    this.trigger(this.state);
+  },
+
+  updateOrganization(org: Organization) {
+    this.state = {
+      ...this.state,
```
Comment (markstory) [inline src/sentry/static/sentry/app/stores/releaseStore.tsx:79]: Should the release & deploy state be reset when the organization changes? This action is called from the OrganizationContext to set the organization. It looks like changing orgs would no longer reset the store.

Comment (leedongwei) [reply src/sentry/static/sentry/app/stores/releaseStore.tsx]: Oh yea, you're right. Fixed.

**Review by markstory — APPROVED** [review]

**Pull request merged**